iml-gui/crate/src/page/filesystem.rs

```diff
@@ -1,145 +1,237 @@
 use crate::{
-    components::{alert_indicator, lock_indicator, pie_chart, table as T, Placement},
+    components::{action_dropdown, alert_indicator, lock_indicator, pie_chart, table as T, Placement},
     extensions::MergeAttrs,
     extract_id,
     generated::css_classes::C,
     route::RouteId,
-    Model, Route,
+    GMsg, Route,
 };
-use im::HashMap;
 use iml_wire_types::{
-    warp_drive::{ArcCache, ArcValuesExt},
-    Filesystem, ResourceUri, Target, TargetConfParam, TargetKind, ToCompositeId, VolumeOrResourceUri,
+    warp_drive::{ArcCache, ArcValuesExt, Locks},
+    Filesystem, Target, TargetConfParam, TargetKind, ToCompositeId, VolumeOrResourceUri,
 };
 use number_formatter as NF;
 use seed::{prelude::*, *};
-use std::sync::Arc;
-
-pub(crate) fn view<I>(model: &Model, f: &Filesystem) -> Node<I> {
-    let (mgt, mut mdt, mut ost) = model.records.target.values().filter(|t| is_fs_target(f, t)).fold(
-        (vec![], vec![], vec![]),
-        |(mut mgt, mut mdt, mut ost), t| {
-            match t.kind {
-                TargetKind::Mgt => mgt.push(t),
-                TargetKind::Mdt => mdt.push(t),
-                TargetKind::Ost => ost.push(t),
+use std::collections::{HashMap, HashSet};
+
+pub struct Row {
+    dropdown: action_dropdown::Model,
+}
+
+#[derive(Default)]
+pub struct Model {
+    pub id: u32,
+    pub mdts: Vec<Target<TargetConfParam>>,
+    pub mgt: Vec<Target<TargetConfParam>>,
+    pub osts: Vec<Target<TargetConfParam>>,
+    pub rows: HashMap<u32, Row>,
+}
+
+#[derive(Clone)]
+pub enum Msg {
+    ActionDropdown(Box<action_dropdown::IdMsg>),
+    // AddTarget(Target<TargetConfParam>),
+    // RemoveTarget(Target<TargetConfParam>),
+    SetTargets(Vec<Target<TargetConfParam>>),
+    WindowClick,
+}
+
+pub fn init(cache: &ArcCache, orders: &mut impl Orders<Msg, GMsg>) {
+    orders.send_msg(Msg::SetTargets(cache.target.arc_values().cloned().collect()));
+}
+
+pub fn update(msg: Msg, cache: &ArcCache, model: &mut Model, orders: &mut impl Orders<Msg, GMsg>) {
+    match msg {
+        Msg::ActionDropdown(x) => {
+            let action_dropdown::IdMsg(id, msg) = *x;
+
+            if let Some(x) = model.rows.get_mut(&id) {
+                action_dropdown::update(
+                    action_dropdown::IdMsg(id, msg),
+                    cache,
+                    &mut x.dropdown,
+                    &mut orders.proxy(|x| Msg::ActionDropdown(Box::new(x))),
+                );
+            }
+        }
+        Msg::SetTargets(xs) => {
+            let old_keys: HashSet<u32> = model.rows.keys().copied().collect();
+            let new_keys: HashSet<u32> = xs.iter().map(|x| x.id).collect();
+
+            let to_remove = &old_keys - &new_keys;
+            let to_add = &new_keys - &old_keys;
+
+            for x in to_remove {
+                model.rows.remove(&x);
```

Comment:
It looks like we are passing all the targets (for all filesystems). Should we pre-filter them first like it is done below (`filter(|t| is_fs_target(model.id, t))`)?